File: mobile/ios/Mesh/Core/MeshDeepRecoveryService.swift
import Combine
import Foundation

/// Privacy deep recovery — continues after Wallet Privacy is closed.
@MainActor
final class MeshDeepRecoveryService: ObservableObject {
    static let shared = MeshDeepRecoveryService()

    @Published private(set) var isRunning = false
    @Published private(set) var progressChecked = 0
    @Published private(set) var progressTotal = Int(MeshPrivacyStore.deepRecoveryScanAddressCount)
    @Published private(set) var isTransferring = false
    @Published private(set) var statusMessage: String?
    @Published private(set) var errorMessage: String?

    private var recoveryTask: Task<Void, Never>?

    private init() {}

    var progressFraction: Double {
        guard progressTotal > 0 else { return 0 }
        return min(1, Double(progressChecked) / Double(progressTotal))
    }

    func start(walletID: String? = MeshWalletRegistry.activeWalletID) {
        guard !isRunning else { return }

        isRunning = true
        errorMessage = nil
        statusMessage = nil
        isTransferring = false
        progressChecked = 0
        progressTotal = Int(MeshPrivacyStore.deepRecoveryScanAddressCount)

        recoveryTask = Task(priority: .userInitiated) { @MainActor in
            await run(walletID: walletID)
        }
    }

    func cancel() {
        recoveryTask?.cancel()
        recoveryTask = nil
        isRunning = false
        isTransferring = false
    }

    private func run(walletID: String?) async {
        defer {
            isRunning = false
            isTransferring = false
            recoveryTask = nil
        }

        do {
            let count = try await MeshPrivacyService.recoverDeepFundsToMainWallet(
                walletID: walletID
            ) { [weak self] progress in
                Task { @MainActor in
                    self?.apply(progress)
                }
            }
            statusMessage = L10n.Send.deepRecoveryDone(count)
            errorMessage = nil
            NotificationCenter.default.post(name: .meshWalletBalancesShouldRefresh, object: nil)
            try? await Task.sleep(nanoseconds: 2_500_000_000)
            statusMessage = nil
        } catch is CancellationError {
            errorMessage = nil
        } catch {
            if MeshPrivacyService.isRateLimitError(error) {
                errorMessage = L10n.Send.deepRecoveryRateLimited
            } else {
                errorMessage = SendErrorPresenter.message(for: error)
            }
            statusMessage = nil
        }
    }

    private func apply(_ progress: MeshPrivacyService.DeepRecoveryProgress) {
        switch progress {
        case let .scanning(checked, total):
            isTransferring = false
            progressChecked = checked
            progressTotal = total
        case let .transferring(current, total):
            isTransferring = true
            progressChecked = current
            progressTotal = total
        }
    }
}

File: mobile/android/app/src/main/java/com/mesh/wallet/core/privacy/MeshDeepRecoveryService.kt
package com.mesh.wallet.core.privacy

import com.mesh.wallet.core.l10n.L10n
import com.mesh.wallet.core.network.SendErrorPresenter
import com.mesh.wallet.core.session.WalletSession
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class MeshDeepRecoveryService(private val session: WalletSession) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private val _isRunning = MutableStateFlow(false)
    val isRunning: StateFlow<Boolean> = _isRunning.asStateFlow()

    private val _progressChecked = MutableStateFlow(0)
    val progressChecked: StateFlow<Int> = _progressChecked.asStateFlow()

    private val _progressTotal = MutableStateFlow(PrivacyStore.DEEP_RECOVERY_SCAN_COUNT)
    val progressTotal: StateFlow<Int> = _progressTotal.asStateFlow()

    private val _isTransferring = MutableStateFlow(false)
    val isTransferring: StateFlow<Boolean> = _isTransferring.asStateFlow()

    private val _statusMessage = MutableStateFlow<String?>(null)
    val statusMessage: StateFlow<String?> = _statusMessage.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    val progressFraction: Float
        get() = if (_progressTotal.value > 0) {
            (_progressChecked.value.toFloat() / _progressTotal.value).coerceAtMost(1f)
        } else 0f

    fun start(walletId: String? = session.registry.activeWalletId) {
        if (_isRunning.value) return
        val id = walletId ?: return
        _isRunning.value = true
        _errorMessage.value = null
        _statusMessage.value = null
        _isTransferring.value = false
        _progressChecked.value = 0
        _progressTotal.value = PrivacyStore.DEEP_RECOVERY_SCAN_COUNT

        scope.launch {
            try {
                val count = session.privacyService.recoverDeepFundsToMainWallet(id) { progress ->
                    when (progress) {
                        is com.mesh.wallet.data.privacy.MeshPrivacyService.DeepRecoveryProgress.Scanning -> {
                            _isTransferring.value = false
                            _progressChecked.value = progress.checked
                            _progressTotal.value = progress.total
                        }
                        is com.mesh.wallet.data.privacy.MeshPrivacyService.DeepRecoveryProgress.Transferring -> {
                            _isTransferring.value = true
                            _progressChecked.value = progress.current
                            _progressTotal.value = progress.total
                        }
                    }
                }
                _statusMessage.value = "Recovered $count address(es)"
                session.registry.activeWalletId?.let { session.backgroundSendService.refreshFeeStatus(it) }
                delay(2_500)
                _statusMessage.value = null
            } catch (e: CancellationException) {
                _errorMessage.value = null
            } catch (e: Exception) {
                _errorMessage.value = SendErrorPresenter.messageFor(e)
                _statusMessage.value = null
            } finally {
                _isRunning.value = false
                _isTransferring.value = false
            }
        }
    }

    fun cancel() {
        _isRunning.value = false
        _isTransferring.value = false
    }
}
